### PR TITLE
[ShapeInference] FFT based - revision and tests (part 1)

### PR DESCRIPTION
**Pull request by mitruska**
### Details:
[Part I] DFT, IDFT, RDFT, IRDFT shape inference revision:
 - additional tests for labels
 - reusage of common helpers between shape infer
 - common type prop tests for DFT and IDFT
 
 \* All of the ops already had shape infer function, and they are registered for CPU
 
 - No logical improvements in this PR, they will be proposed separately to reduce number of changes
### Tickets:
 - Part of 97219

### CODE REVIEW

File: src/core/shape_inference/include/fft_common_validation.hpp [hunk 1]
```diff
@@ -91,9 +91,9 @@ void validate_axes(const ov::op::util::FFTBase* op,
     for (int64_t& axis : axes) {
```
Comment (praasz) [inline src/core/shape_inference/include/fft_common_validation.hpp:91]: As part 2 (improvments): Check if this loop can be replaced by `normalize_axes` or use `normalize_axis` from validation_util.

File: src/core/shape_inference/include/irdft_shape_inference.hpp [hunk 2]
```diff
@@ -26,11 +26,11 @@ std::vector<TRShape> shape_infer(const IRDFT* op,
...
 
-    util::rfft_common_validation::shape_validation(op,
-                                                   input_shapes,
-                                                   *axes,
-                                                   axes_are_known,
-                                                   util::rfft_common_validation::RFFTKind::Inverse);
+    util::fft_common_validation::shape_validation(op,
+                                                  input_shapes,
+                                                  *axes,
+                                                  axes_are_known,
```
Comment (praasz) [inline src/core/shape_inference/include/irdft_shape_inference.hpp:32]: As part 2 (improvments)?: The axes_are_known paramter I think can be removed by wrap this call in if (axes != nullptr) or do this check inside function.

File: src/core/tests/type_prop/fft_base_complex.cpp [hunk 3]
```diff
@@ -0,0 +1,514 @@
...
+    PartialShape ref_output_shape;
+    std::vector<int64_t> axes;
+    std::vector<int64_t> signal_size;
+    std::vector<ov::label_t> expected_labels;
+};
+
+template <class TOp>
+class FFTConstantAxesAndConstantSignalSizeTest : public TypePropOpTest<TOp> {
+public:
+    std::vector<FFTConstantAxesAndConstantSignalSizeTestParams> test_params{
```
Comment (praasz) [inline src/core/tests/type_prop/fft_base_complex.cpp:33]: Consider make this vector `static const` or `const` at least?

File: src/core/tests/type_prop/fft_base_complex.cpp [hunk 4]
```diff
@@ -0,0 +1,514 @@
...
+struct NonConstantAxesTestParams {
+    PartialShape input_shape;
+    PartialShape axes_shape;
+    PartialShape ref_output_shape;
+};
+
+template <class TOp>
+class FFTNonConstantAxesTest : public TypePropOpTest<TOp> {
+public:
+    std::vector<NonConstantAxesTestParams> test_params{
```
Comment (praasz) [inline src/core/tests/type_prop/fft_base_complex.cpp:216]: Make it `static const`?
Check similar usage.

File: src/core/tests/type_prop/fft_base_complex.cpp [hunk 5]
```diff
@@ -0,0 +1,514 @@
...
+
+template <class TOp>
+class FFTInvalidInput : public TypePropOpTest<TOp> {};
+
+TYPED_TEST_SUITE_P(FFTInvalidInput);
+
+TYPED_TEST_P(FFTInvalidInput, invalid_input_data) {
+    auto axes = op::v0::Constant::create(element::i64, Shape{2}, {0, 1});
+
+    try {
```
Comment (praasz) [inline src/core/tests/type_prop/fft_base_complex.cpp:398]: Consider refactor to use `OV_EXPECT_THROW` macro instead `try...catch`?
Check similar usages in modified test files.

File: src/core/tests/type_prop/fft_base_complex.cpp [hunk 6]
```diff
@@ -0,0 +1,514 @@
...
+TYPED_TEST_P(FFTDynamicTypes, dynamic_types) {
+    const auto input_shape = PartialShape{2, 180, 180, 2};
+    const auto axes_shape = PartialShape::dynamic();
+    const auto signal_size_shape = PartialShape::dynamic();
+    const auto ref_output_shape = PartialShape{Dimension::dynamic(), Dimension::dynamic(), Dimension::dynamic(), 2};
+
+    auto data = std::make_shared<op::v0::Parameter>(element::dynamic, input_shape);
+    auto axes_input = std::make_shared<op::v0::Parameter>(element::dynamic, axes_shape);
+    auto signal_size_input = std::make_shared<op::v0::Parameter>(element::dynamic, signal_size_shape);
+    auto dft = std::make_shared<TypeParam>(data, axes_input, signal_size_input);
```
Comment (praasz) [inline src/core/tests/type_prop/fft_base_complex.cpp:505]:
```suggestion
    auto dft = this->make_op(data, axes_input, signal_size_input);
```
?
Check similar cases in modified tests and/or test fixtures.

File: src/core/shape_inference/include/fft_common_validation.hpp [hunk 7]
```diff
@@ -104,23 +104,23 @@ void validate_axes(const ov::op::util::FFTBase* op,
         axes_set.insert(static_cast<size_t>(axis));
     }
 
-    NODE_VALIDATION_CHECK(op, axes.size() == axes_set.size(), "(I)RDFT op axes must be unique.");
+    NODE_VALIDATION_CHECK(op, axes.size() == axes_set.size(), "Each axis must be unique.");
 }
 
 template <class T>
 void validate_signal_size(const ov::op::util::FFTBase* op, const T& axes_shape, const T& signal_size_shape) {
     NODE_VALIDATION_CHECK(op,
```
Comment (praasz) [inline src/core/shape_inference/include/fft_common_validation.hpp:112]: As exception message has been modified, consider to use `NODE_SHAPE_INFER_CHECK` and reduce message strings by e.g. remove `signal_size_shape`.
Comment applies to all modified `NODE_VALIDATION_CHECK`

File: src/core/shape_inference/include/fft_common_validation.hpp [hunk 8]
```diff
@@ -62,13 +62,13 @@ void validate_axes(const ov::op::util::FFTBase* op,
                    std::vector<int64_t>& axes,
                    size_t input_rank,
```
Comment (praasz) [inline src/core/shape_inference/include/fft_common_validation.hpp:63]: As future improvement consider change it to be int64_t as it is casted inside function. Maybe pass `Rank` object or for shape inference validation there will be required to pass all input shapes and rank can be check internally etc.

File: src/core/shape_inference/include/fft_common_validation.hpp [hunk 9]
```diff
@@ -10,23 +10,23 @@
 namespace ov {
 namespace op {
 namespace util {
-namespace rfft_common_validation {
-enum class RFFTKind { Forward, Inverse };
+namespace fft_common_validation {
+enum class FFTKind { RealInput, ComplexInput };
```
Comment (praasz) [inline src/core/shape_inference/include/fft_common_validation.hpp:14]: Future improvement, use c-style `enum` or some `constexpr` values for this FFT kind instead? 
Then 
```const size_t min_rank = (fft_kind == FFTKind::RealInput) ? 1 : 2;``` 
will be not required as FFT kind will have correct value.

File: src/core/tests/type_prop/fft_base_complex.cpp (hunk 4 above)
Comment (mitruska) [reply src/core/tests/type_prop/fft_base_complex.cpp]: Updated to `const`, `static const` causes compilation error.

File: src/core/tests/type_prop/fft_base_complex.cpp (hunk 5 above)
Comment (mitruska) [reply src/core/tests/type_prop/fft_base_complex.cpp]: Updated to use `OV_EXPECT_THROW`

File: src/core/tests/type_prop/fft_base_complex.cpp (hunk 6 above)
Comment (mitruska) [reply src/core/tests/type_prop/fft_base_complex.cpp]: Usage of <TypeParam> is not wrong, but okay, updated;)

File: src/core/shape_inference/include/fft_common_validation.hpp (hunk 7 above)
Comment (mitruska) [reply src/core/shape_inference/include/fft_common_validation.hpp]: Updated to pass `input_shapes` to the validation helpers and use `NODE_SHAPE_INFER_CHECK`.

**Comment by mitruska** [comment]
The comments related to this part are applied, future improvements will be considered separately.

**Review by praasz — APPROVED** [review]

File: src/core/shape_inference/include/fft_common_validation.hpp (hunk 9 above)
Comment (mitruska) [reply src/core/shape_inference/include/fft_common_validation.hpp:14]: Added as a part of https://github.com/openvinotoolkit/openvino/pull/19463.

File: src/core/shape_inference/include/fft_common_validation.hpp (hunk 8 above)
Comment (mitruska) [reply src/core/shape_inference/include/fft_common_validation.hpp:58]: Added as a part of https://github.com/openvinotoolkit/openvino/pull/19463.

**Pull request merged**